File: src/main/java/kr/co/bookand/backend/account/service/AuthService.java
package kr.co.bookand.backend.account.service;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.jsonwebtoken.Claims;
import io.jsonwebtoken.ExpiredJwtException;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.MalformedJwtException;
import io.jsonwebtoken.security.SignatureException;
import kr.co.bookand.backend.account.domain.Account;
import kr.co.bookand.backend.account.domain.Role;
import kr.co.bookand.backend.account.domain.SocialType;
import kr.co.bookand.backend.account.domain.dto.AccountDto;
import kr.co.bookand.backend.account.domain.dto.AppleDto;
import kr.co.bookand.backend.account.domain.dto.AuthDto;
import kr.co.bookand.backend.account.domain.dto.TokenDto;
import kr.co.bookand.backend.account.exception.*;
import kr.co.bookand.backend.account.repository.AccountRepository;
import kr.co.bookand.backend.account.util.SecurityUtil;
import kr.co.bookand.backend.bookmark.domain.Bookmark;
import kr.co.bookand.backend.bookmark.domain.BookmarkType;
import kr.co.bookand.backend.bookmark.repository.BookmarkRepository;
import kr.co.bookand.backend.common.service.RestTemplateService;
import kr.co.bookand.backend.common.exception.ErrorCode;
import kr.co.bookand.backend.common.domain.Message;
import kr.co.bookand.backend.config.jwt.exception.JwtException;
import kr.co.bookand.backend.config.jwt.RefreshToken;
import kr.co.bookand.backend.config.jwt.RefreshTokenRepository;
import kr.co.bookand.backend.config.jwt.TokenFactory;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.http.*;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.MultiValueMap;


import java.math.BigInteger;
import java.nio.charset.StandardCharsets;
import java.security.KeyFactory;
import java.security.NoSuchAlgorithmException;
import java.security.PublicKey;
import java.security.spec.InvalidKeySpecException;
import java.security.spec.RSAPublicKeySpec;
import java.util.*;

import static kr.co.bookand.backend.account.domain.dto.AuthDto.*;
import static kr.co.bookand.backend.account.domain.dto.TokenDto.*;

@Slf4j
@Service
@RequiredArgsConstructor
public class AuthService {

    private final RefreshTokenRepository refreshTokenRepository;
    private final TokenFactory tokenFactory;
    private final AuthenticationManagerBuilder authenticationManagerBuilder;
    private final AccountRepository accountRepository;
    private final BookmarkRepository bookmarkRepository;
    private final RestTemplateService<MultiValueMap<String, String>> apiService;
    private final PasswordEncoder passwordEncoder;

    private static final String INIT_BOOKMARK_FOLDER_NAME = "모아보기";

    ParameterizedTypeReference<Map<String, Object>> RESPONSE_TYPE = new ParameterizedTypeReference<>() {
    };

    @Value("${bookand.suffix}")
    private String suffix;

    @Transactional
    public LoginResponse socialAccess(AuthRequest authRequestDto) {
        String userId = getSocialIdWithAccessToken(authRequestDto).getUserId();
        String providerEmail = getSocialIdWithAccessToken(authRequestDto).getEmail();
        authRequestDto.insertId(userId);
        String email = authRequestDto.extraEmail();
        Optional<Account> account = accountRepository.findByEmail(email);
        SocialType socialType = authRequestDto.getSocialType();

        if (account.isPresent()) {
            // 로그인
            TokenDto tokenDto = login(account.get().toAccountRequestDto(suffix).toLoginRequest());
            TokenResponse tokenResponse = tokenDto.toTokenDto();
            return LoginResponse.builder().tokenResponse(tokenResponse).httpStatus(HttpStatus.OK).build();

        } else {
            MiddleAccount middleAccount = MiddleAccount.builder()
                    .email(email)
                    .socialType(socialType)
                    .providerEmail(providerEmail)
                    .build();
            // 회원가입
            SignDto signTokenDto = tokenFactory.createSignTokenDto(middleAccount);
            return LoginResponse.builder().tokenResponse(signTokenDto).httpStatus(HttpStatus.NOT_FOUND).build();
        }
    }


    public TokenDto login(AccountDto.LoginRequest loginRequest) {
        String email = loginRequest.getEmail();
        accountRepository.findByEmail(email).orElseThrow(() -> new AccountException(ErrorCode.NOT_FOUND_MEMBER, email));
        return getTokenDto(loginRequest);
    }

    public TokenDto loginAdmin(AccountDto.LoginRequest loginRequest) {

        UsernamePasswordAuthenticationToken authenticationToken = loginRequest.toAuthentication();
        Authentication authentication = authenticationManagerBuilder.getObject().authenticate(authenticationToken);
        for (GrantedAuthority grantedAuthority : authentication.getAuthorities()) {
            String authority = grantedAuthority.getAuthority();
            checkRole(Role.ADMIN.name(), authority);
        }

        return getTokenDto(loginRequest);
    }

    public TokenDto loginManager(AccountDto.LoginRequest loginRequest) {
        UsernamePasswordAuthenticationToken authenticationToken = loginRequest.toAuthentication();
        Authentication authentication = authenticationManagerBuilder.getObject().authenticate(authenticationToken);
        for (GrantedAuthority grantedAuthority : authentication.getAuthorities()) {
            String authority = grantedAuthority.getAuthority();
            checkRole(Role.MANAGER.name(), authority);
        }
        return getTokenDto(loginRequest);
    }

    private void checkRole(String role_manager, String authority) {
        if (!role_manager.equals(authority)) {
            throw new AccountException(ErrorCode.NOT_ROLE_MEMBER, authority);
        }
    }

    private TokenDto getTokenDto(AccountDto.LoginRequest loginRequest) {
        UsernamePasswordAuthenticationToken authenticationToken = loginRequest.toAuthentication();
        Authentication authentication = authenticationManagerBuilder.getObject().authenticate(authenticationToken);
        TokenDto tokenDto = tokenFactory.generateTokenDto(authentication);

        // refresh token 저장
        RefreshToken refreshToken = RefreshToken.builder()
                .key(authentication.getName())
                .value(tokenDto.getRefreshToken())
                .build();
        refreshTokenRepository.save(refreshToken);
        return tokenDto;
    }

    @Transactional
    public TokenResponse socialSignUp(SignDto signTokenDto) {
        String signToken = signTokenDto.signToken();
        SigningAccount signKey = tokenFactory.getSignKey(signToken);
        checkSignUp(signKey);
        String nickname = nicknameRandom();
        String email = signKey.email();
        String socialType = signKey.socialType();
        String providerEmail = signKey.providerEmail();
        duplicateEmailAndNickName(email, nickname);
        Account account = signTokenDto.toAccount(email, socialType, providerEmail, passwordEncoder, suffix, nickname);
        Account saveAccount = accountRepository.save(account);

        List<Bookmark> initBookmark = createInitBookmark(saveAccount);
        saveAccount.updateBookmarkList(initBookmark);

        TokenDto tokenDto = login(account.toAccountRequestDto(suffix).toLoginRequest());
        return tokenDto.toTokenDto();
    }

    private List<Bookmark> createInitBookmark(Account saveAccount) {
        Bookmark initBookmarkArticle = Bookmark.builder()
                .account(saveAccount)
                .folderName(INIT_BOOKMARK_FOLDER_NAME)
                .bookmarkType(BookmarkType.ARTICLE)
                .build();
        Bookmark initBookmarkBookStore = Bookmark.builder()
                .account(saveAccount)
                .folderName(INIT_BOOKMARK_FOLDER_NAME)
                .bookmarkType(BookmarkType.BOOKSTORE)
                .build();

        bookmarkRepository.save(initBookmarkArticle);
        bookmarkRepository.save(initBookmarkBookStore);
        return Arrays.asList(initBookmarkArticle, initBookmarkBookStore);
    }

    public void checkSignUp(SigningAccount signKey) {
        if (signKey.email() == null || signKey.socialType() == null || signKey.providerEmail() == null) {
            throw new AccountException(ErrorCode.INVALID_SIGN_TOKEN, signKey);
        }
    }

    private String nicknameRandom() {
        String url = "https://nickname.hwanmoo.kr/?format=json&count=1&max_length=10";
        HttpHeaders headers = new HttpHeaders();
        headers.setAccept(Collections.singletonList(MediaType.APPLICATION_JSON));
        HttpEntity<MultiValueMap<String, String>> request = new HttpEntity<>(headers);
        ResponseEntity<Map<String, Object>> response = apiService.httpEntityPost(url, HttpMethod.GET, request, RESPONSE_TYPE);
        Map<String, Object> stringObjectMap = Objects.requireNonNull(response.getBody());
        return stringObjectMap.get("words").toString().replaceAll("\\[", "").replaceAll("]", "");
    }

    private void duplicateEmailAndNickName(String email, String nickname) {
        if (accountRepository.existsByEmail(email)) {
            throw new AccountException(ErrorCode.EMAIL_DUPLICATION, email);
        }
        if (accountRepository.existsByNickname(nickname)) {
            throw new AccountException(ErrorCode.NICKNAME_DUPLICATION, nickname);
        }
    }


    private ProviderIdAndEmail getSocialIdWithAccessToken(AuthRequest data) {
        SocialType socialType = data.getSocialType();
        if (socialType.equals(SocialType.GOOGLE)) {
            return getGoogleIdAndEmail(data);
        } else if (socialType.equals(SocialType.APPLE)) {
            return getAppleId(data.getAccessToken());
        } else {
            throw new RuntimeException();
        }
    }

    private ProviderIdAndEmail getGoogleIdAndEmail(AuthRequest data) {
        String url = data.getSocialType().getUserInfoUrl();
        HttpMethod method = data.getSocialType().getMethod();
        HttpHeaders headers = setHeaders(data.getAccessToken());
        HttpEntity<MultiValueMap<String, String>> request = new HttpEntity<>(headers);
        ResponseEntity<Map<String, Object>> response = apiService.httpEntityPost(url, method, request, RESPONSE_TYPE);
        String userId = (String) response.getBody().get("sub");
        String providerEmail = (String) response.getBody().get("email");
        return ProviderIdAndEmail.toProviderDto(userId, providerEmail);
    }

    private AppleDto getAppleAuthPublicKey() {
        String url = SocialType.APPLE.getUserInfoUrl();
        HttpMethod method = SocialType.APPLE.getMethod();
        HttpHeaders headers = new HttpHeaders();
        HttpEntity<MultiValueMap<String, String>> request = new HttpEntity<>(headers);
        ResponseEntity<AppleDto> response = apiService.getAppleKeys(url, method, request, AppleDto.class);
        return response.getBody();
    }

    private ProviderIdAndEmail getAppleId(String identityToken) {
        AppleDto appleKeyStorage = getAppleAuthPublicKey();
        try {
            String headerToken = identityToken.substring(0, identityToken.indexOf("."));
            Map<String, String> header = new ObjectMapper().readValue(new String(Base64.getDecoder().decode(headerToken), StandardCharsets.UTF_8), Map.class);
            AppleDto.AppleKey key = appleKeyStorage.getMatchedKeyBy(header.get("kid"), header.get("alg")).orElseThrow();

            byte[] nBytes = Base64.getUrlDecoder().decode(key.n());
            byte[] eBytes = Base64.getUrlDecoder().decode(key.e());

            BigInteger n = new BigInteger(1, nBytes);
            BigInteger e = new BigInteger(1, eBytes);

            RSAPublicKeySpec publicKeySpec = new RSAPublicKeySpec(n, e);
            KeyFactory keyFactory = KeyFactory.getInstance(key.kty());
            PublicKey publicKey = keyFactory.generatePublic(publicKeySpec);

            Claims claims = Jwts.parserBuilder().setSigningKey(publicKey).build().parseClaimsJws(identityToken).getBody();
            String subject = claims.getSubject();
            String email = (String) claims.get("email");
            return ProviderIdAndEmail.toProviderDto(subject, email);

        } catch (JsonProcessingException | NoSuchAlgorithmException | InvalidKeySpecException | SignatureException |
                MalformedJwtException | ExpiredJwtException | IllegalArgumentException e) {
            throw new AccountException(ErrorCode.APPLE_LOGIN_ERROR, e);
        }
    }

    public HttpHeaders setHeaders(String accessToken) {
        HttpHeaders headers = new HttpHeaders();
        headers.set("Authorization", "Bearer " + accessToken);
        headers.setContentType(MediaType.APPLICATION_FORM_URLENCODED);
        headers.setAccept(Collections.singletonList(MediaType.APPLICATION_JSON));
        return headers;
    }

    @Transactional
    public Message logout() {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        String loginAccount = authentication.getName();
        if (accountRepository.findByEmail(loginAccount).isPresent() ||
                refreshTokenRepository.findByKey(authentication.getName()).isPresent()) {
            // 리프레시 토큰 삭제
            RefreshToken refreshToken = refreshTokenRepository.findByKey(authentication.getName()).get();
            refreshTokenRepository.delete(refreshToken);
        } else {
            throw new AccountException(ErrorCode.NOT_FOUND_MEMBER, loginAccount);
        }
        return Message.of("로그아웃 성공");
    }

    @Transactional
    public TokenResponse reissue(TokenRequest tokenRequestDto) {
        if (!tokenFactory.validateToken(tokenRequestDto.getRefreshToken())) {
            throw new JwtException(ErrorCode.JWT_ERROR, "토큰이 유효하지 않습니다.");
        }
        Authentication authentication = tokenFactory.getAuthentication(tokenRequestDto.getRefreshToken());
        RefreshToken refreshToken = refreshTokenRepository.findByKey(authentication.getName())
                .orElseThrow(() -> new JwtException(ErrorCode.NOT_FOUND_REFRESH_TOKEN, ErrorCode.NOT_FOUND_REFRESH_TOKEN.getMessage()));
        reissueRefreshExceptionCheck(refreshToken.getValue(), tokenRequestDto);
        TokenDto tokenDto = tokenFactory.generateTokenDto(authentication);

        // refresh token 저장
        RefreshToken newRefreshToken = RefreshToken.builder()
                .key(authentication.getName())
                .value(tokenDto.getRefreshToken())
                .build();
        refreshTokenRepository.save(newRefreshToken);

        return tokenDto.toTokenDto();
    }

    private void reissueRefreshExceptionCheck(String refreshToken, TokenRequest tokenRequestDto) {
        if (refreshToken == null) {
            throw new JwtException(ErrorCode.NOT_FOUND_REFRESH_TOKEN, ErrorCode.NOT_FOUND_REFRESH_TOKEN.getMessage());
        }
        if (!refreshToken.equals(tokenRequestDto.getRefreshToken())) {
            throw new JwtException(ErrorCode.NOT_MATCH_REFRESH_TOKEN, ErrorCode.NOT_MATCH_REFRESH_TOKEN.getMessage());
        }
    }

    public TokenResponse adminLogin(AccountDto.LoginRequest loginRequestDto) {
        TokenDto tokenDto = loginAdmin(loginRequestDto);
        return tokenDto.toTokenDto();
    }

    public TokenResponse managerLogin(AccountDto.LoginRequest loginRequestDto) {
        TokenDto tokenDto = loginManager(loginRequestDto);
        return tokenDto.toTokenDto();
    }

    public Message createManager(AccountDto.ManagerInfo createRequestDto) {
        Account admin = accountRepository.findByEmail(SecurityUtil.getCurrentAccountEmail())
                .orElseThrow(() -> new AccountException(ErrorCode.NOT_FOUND_MEMBER, SecurityUtil.getCurrentAccountEmail()));
        admin.getRole().checkAdmin();
        duplicateEmailAndNickName(createRequestDto.email(), createRequestDto.nickname());

        AuthDto.MiddleAccount middleAccount = AuthDto.MiddleAccount.builder()
                .email(createRequestDto.email())
                .socialType(SocialType.GOOGLE)
                .providerEmail("providerEmail")
                .build();
        Account account = middleAccount.toManager(passwordEncoder, createRequestDto.email(), createRequestDto.password(), createRequestDto.nickname());
        Account saveAccount = accountRepository.save(account);

        List<Bookmark> initBookmark = createInitBookmark(saveAccount);
        saveAccount.updateBookmarkList(initBookmark);
        return Message.of("생성 완료");
    }

}

File: src/main/kotlin/kr/co/bookand/backend/account/service/AuthService.kt
package kr.co.bookand.backend.account.service

import com.fasterxml.jackson.databind.ObjectMapper
import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jwts
import kr.co.bookand.backend.account.dto.*
import kr.co.bookand.backend.account.model.*
import kr.co.bookand.backend.account.repository.AccountRepository
import kr.co.bookand.backend.account.repository.SuspendedAccountRepository
import kr.co.bookand.backend.bookmark.model.Bookmark
import kr.co.bookand.backend.bookmark.model.BookmarkType
import kr.co.bookand.backend.bookmark.repository.BookmarkRepository
import kr.co.bookand.backend.common.ErrorCode
import kr.co.bookand.backend.common.RestTemplateService
import kr.co.bookand.backend.common.model.MessageResponse
import kr.co.bookand.backend.common.exception.BookandException
import kr.co.bookand.backend.config.jwt.*
import org.springframework.beans.factory.annotation.Value
import org.springframework.core.ParameterizedTypeReference
import org.springframework.http.*
import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.util.MultiValueMap
import java.math.BigInteger
import java.nio.charset.StandardCharsets
import java.security.KeyFactory
import java.security.PublicKey
import java.security.spec.RSAPublicKeySpec
import java.time.LocalDateTime
import java.util.*


@Service
@Transactional(readOnly = true)
class AuthService(
    private val jwtProvider: JwtProvider,
    private val authenticationManagerBuilder: AuthenticationManagerBuilder,
    private val apiService: RestTemplateService<MultiValueMap<String, String>>,
    private val passwordEncoder: PasswordEncoder,
    private val refreshTokenRepository: RefreshTokenRepository,
    private val accountRepository: AccountRepository,
    private val bookmarkRepository: BookmarkRepository,
    private val suspendedAccountRepository: SuspendedAccountRepository
) {
    val responseType: ParameterizedTypeReference<Map<String, Any>> =
        object : ParameterizedTypeReference<Map<String, Any>>() {}

    @Value("\${bookand.suffix}")
    private lateinit var suffix: String

    @Transactional
    fun socialAccess(authRequest: AuthRequest): LoginResponse {
        val socialIdWithAccessToken = getSocialIdWithAccessToken(authRequest)
        authRequest.insertId(socialIdWithAccessToken.userId)
        val email = authRequest.extraEmail()
        val providerEmail = socialIdWithAccessToken.email
        val existAccount = accountRepository.findByEmail(email)
        val socialType = authRequest.getSocialType()

        if (existAccount != null) {
            // 로그인
            val tokenResponse =
                login(AccountInfoResponse(account = existAccount).toLoginRequest(suffix))
            return LoginResponse(
                tokenResponse = tokenResponse,
                httpStatus = HttpStatus.OK
            )
        } else {
            val middleAccount = MiddleAccount(
                email = email,
                providerEmail = providerEmail,
                socialType = socialType,
            )
            // 회원가입
            val signTokenResponse = jwtProvider.createSignTokenDto(middleAccount)
            return LoginResponse(
                tokenResponse = signTokenResponse,
                httpStatus = HttpStatus.NOT_FOUND
            )
        }
    }

    @Transactional
    fun socialSignUp(signTokenRequest: SignTokenRequest): TokenResponse {
        val signKey = jwtProvider.getSignKey(signTokenRequest)
        checkSignUp(signKey)
        val nickname = nicknameRandom()
        duplicateEmailAndNickName(signKey.email, nickname)
        val account = Account(
            email = signKey.email,
            nickname = nickname,
            password = passwordEncoder.encode(signKey.email + suffix),
            provider = signKey.socialType,
            providerEmail = signKey.providerEmail,
            role = Role.USER,
            accountStatus = AccountStatus.NORMAL,
            signUpDate = LocalDateTime.now(),
        )
        val saveAccount = accountRepository.save(account)
        saveAccount.updateLastLoginDate()

        val initBookmark = createInitBookmark(saveAccount)
        saveAccount.updateBookmarks(initBookmark)

        val tokenResponse = login(AccountInfoResponse(account = saveAccount).toLoginRequest(suffix))

        return TokenResponse(
            accessToken = tokenResponse.accessToken,
            refreshToken = tokenResponse.refreshToken
        )
    }

    fun login(loginRequest: LoginRequest): TokenResponse {
        val email = loginRequest.email
        val account = getAccountByEmail(email)
        checkSuspended(account)
        account.updateLastLoginDate()
        return getTokenDto(loginRequest)
    }

    @Transactional
    fun checkSuspended(account: Account) {
        val suspendedAccount = getSuspendedAccount(account) ?: return
        val isAccountExpired = LocalDateTime.now().isAfter(suspendedAccount.endedSuspendedAt)
        when (account.accountStatus) {
            AccountStatus.SUSPENDED -> {
                if (isAccountExpired) {
                    account.updateAccountStatus(AccountStatus.NORMAL)
                } else {
                    throw BookandException(ErrorCode.SUSPENDED_ACCOUNT)
                }
            }

            AccountStatus.DELETED -> throw BookandException(ErrorCode.DELETED_ACCOUNT)
            else -> return
        }
    }

    private fun checkRole(roleManager: String, authority: String) {
        if (roleManager != authority) {
            throw BookandException(ErrorCode.ROLE_ACCESS_ERROR)
        }
    }

    fun basicLogin(loginRequest: LoginRequest, requiredRole: Role): TokenResponse {
        val authenticationToken = loginRequest.toAuthenticationToken()
        val authentication = authenticationManagerBuilder.getObject().authenticate(authenticationToken)
        for (grantedAuthority in authentication.authorities) {
            val authority = grantedAuthority.authority
            checkRole(requiredRole.name, authority)
        }
        return getTokenDto(loginRequest)
    }

    fun adminLogin(loginRequestDto: LoginRequest): TokenResponse {
        return basicLogin(loginRequestDto, Role.ADMIN)
    }

    fun managerLogin(loginRequestDto: LoginRequest): TokenResponse {
        return basicLogin(loginRequestDto, Role.MANAGER)
    }


    @Transactional
    fun createManager(account: Account, managerInfoRequest: ManagerInfoRequest): MessageResponse {
        val admin = account
        admin.role.checkAdmin()
        duplicateEmailAndNickName(managerInfoRequest.email, managerInfoRequest.nickname)
        val account = Account(
            email = managerInfoRequest.email,
            password = passwordEncoder.encode(managerInfoRequest.password),
            nickname = managerInfoRequest.nickname,
            role = Role.MANAGER,
            provider = SocialType.GOOGLE.name,
            providerEmail = "providerEmail",
            accountStatus = AccountStatus.NORMAL
        )
        val saveAccount = accountRepository.save(account)
        saveAccount.updateBookmarks(createInitBookmark(account))
        return MessageResponse(result = "생성 완료", 200)
    }


    @Transactional
    fun logout(): MessageResponse {
        val authentication = SecurityContextHolder.getContext().authentication
        val loginAccount = authentication.name
        if (accountRepository.existsByEmail(loginAccount) || refreshTokenRepository.existsByKey(authentication.name)) {
            val refreshToken = refreshTokenRepository.findByKey(authentication.name)
                ?: throw BookandException(ErrorCode.NOT_FOUND_REFRESH_TOKEN)
            refreshTokenRepository.delete(refreshToken)
        } else {
            throw BookandException(ErrorCode.NOT_FOUND_MEMBER)
        }
        return MessageResponse(result = "로그아웃 성공", 200)
    }

    private fun createInitBookmark(saveAccount: Account): MutableList<Bookmark> {
        return getBookmarks(saveAccount, INIT_BOOKMARK_FOLDER_NAME, bookmarkRepository)
    }

    @Transactional
    fun getBookmarks(
        saveAccount: Account,
        initBookmarkFolderName: String,
        bookmarkRepository: BookmarkRepository
    ): MutableList<Bookmark> {
        val initBookmarkArticle = Bookmark(
            account = saveAccount,
            folderName = initBookmarkFolderName,
            bookmarkType = BookmarkType.ARTICLE
        )
        val initBookmarkBookStore = Bookmark(
            account = saveAccount,
            folderName = initBookmarkFolderName,
            bookmarkType = BookmarkType.BOOKSTORE
        )

        bookmarkRepository.save(initBookmarkArticle)
        bookmarkRepository.save(initBookmarkBookStore)
        return mutableListOf(initBookmarkArticle, initBookmarkBookStore)
    }


    private fun duplicateEmailAndNickName(email: String, nickname: String) {
        if (accountRepository.existsByEmail(email)) {
            throw BookandException(ErrorCode.EMAIL_DUPLICATION)
        }
        if (accountRepository.existsByNickname(nickname)) {
            throw BookandException(ErrorCode.NICKNAME_DUPLICATION)
        }
    }


    fun checkSignUp(signKey: SigningAccount) {
        if (signKey.email == null || signKey.socialType == null || signKey.providerEmail == null) {
            throw BookandException(ErrorCode.INVALID_SIGN_TOKEN)
        }
    }

    @Transactional
    fun reissue(tokenRequestDto: TokenRequest): TokenResponse {
        if (!jwtProvider.validateToken(tokenRequestDto.refreshToken)) {
            throw BookandException(ErrorCode.JWT_ERROR)
        }
        val authentication = jwtProvider.getAuthentication(tokenRequestDto.refreshToken)
        if (authentication != null) {
            println("authentication.name = ${authentication.name}")
        }
        val refreshToken = authentication?.let { refreshTokenRepository.findByKey(it.name) }
            ?: throw BookandException(ErrorCode.NOT_FOUND_REFRESH_TOKEN)

        checkSuspended(refreshToken.account)

        // 로그인 접근 시간 업데이트
        refreshToken.account.updateLastLoginDate()

        reissueRefreshExceptionCheck(refreshToken.value, tokenRequestDto)
        val tokenDto = jwtProvider.generateToken(authentication)

        // refresh token 업데이트
        refreshToken.updateValue(tokenDto.refreshToken)

        return TokenResponse(
            accessToken = tokenDto.accessToken,
            refreshToken = tokenDto.refreshToken
        )
    }

    private fun reissueRefreshExceptionCheck(refreshToken: String?, tokenRequestDto: TokenRequest) {
        if (refreshToken == null) {
            throw BookandException(ErrorCode.NOT_FOUND_REFRESH_TOKEN)
        }
        if (refreshToken != tokenRequestDto.refreshToken) {
            throw BookandException(ErrorCode.NOT_MATCH_REFRESH_TOKEN)
        }
    }


    fun nicknameRandom(): String {
        val url = "https://nickname.hwanmoo.kr/?format=json&count=1&max_length=10"
        val headers = HttpHeaders()
        headers.accept = listOf(MediaType.APPLICATION_JSON)
        val request = HttpEntity<MultiValueMap<String, String>>(headers)
        val response: ResponseEntity<Map<String, Any>> =
            apiService.httpEntityPost(url, HttpMethod.GET, request, responseType)
        val stringObjectMap = response.body ?: throw RuntimeException("Failed to retrieve nickname")
        return stringObjectMap["words"].toString().replace("[", "").replace("]", "")
    }


    fun getTokenDto(loginRequest: LoginRequest): TokenResponse {
        val authentication = authenticationManagerBuilder.getObject().authenticate(loginRequest.toAuthenticationToken())
        val tokenDto = jwtProvider.generateToken(authentication)
        val account = getAccountByEmail(loginRequest.email)
        val refreshToken = RefreshToken(
            key = loginRequest.email,
            value = tokenDto.refreshToken,
            account = account
        )
        refreshTokenRepository.save(refreshToken)
        return tokenDto
    }


    fun getSocialIdWithAccessToken(data: AuthRequest): ProviderIdAndEmail {
        return when (data.socialType) {
            SocialType.GOOGLE.name -> getGoogleIdAndEmail(data)
            SocialType.APPLE.name -> getAppleId(data.accessToken)
            else -> throw BookandException(ErrorCode.NOT_FOUND_SOCIAL_TYPE)
        }
    }

    private fun getGoogleIdAndEmail(data: AuthRequest): ProviderIdAndEmail {
        val url = data.getSocialType().userInfoUrl
        val method = data.getSocialType().method
        val headers = setHeaders(data.accessToken)
        val request = HttpEntity<MultiValueMap<String, String>>(headers)
        val response: ResponseEntity<Map<String, Any>> = apiService.httpEntityPost(url, method, request, responseType)
        val userId = response.body?.get("sub") as String?
        val providerEmail = response.body?.get("email") as String?
        return ProviderIdAndEmail(userId, providerEmail)
    }

    private fun getAppleAuthPublicKey(): AppleDto {
        val url = SocialType.APPLE.userInfoUrl
        val method = SocialType.APPLE.method
        val headers = HttpHeaders()
        val request = HttpEntity<MultiValueMap<String, String>>(headers)
        val response: ResponseEntity<AppleDto> =
            apiService.getAppleKeys(url, method, request, AppleDto::class.java)
        return response.body
    }

    private fun getAppleId(identityToken: String): ProviderIdAndEmail {
        val appleKeyStorage = getAppleAuthPublicKey()
        try {
            val headerToken = identityToken.substring(0, identityToken.indexOf("."))
            val header: Map<String, String> = ObjectMapper().readValue(
                String(Base64.getDecoder().decode(headerToken), StandardCharsets.UTF_8),
                Map::class.java
            ) as Map<String, String>

            val key: AppleDto.AppleKey =
                appleKeyStorage.getMatchedKeyBy(header["kid"], header["alg"]).orElseThrow()

            val nBytes = Base64.getUrlDecoder().decode(key.n)
            val eBytes = Base64.getUrlDecoder().decode(key.e)

            val n = BigInteger(1, nBytes)
            val e = BigInteger(1, eBytes)

            val publicKeySpec = RSAPublicKeySpec(n, e)
            val keyFactory = KeyFactory.getInstance(key.kty)
            val publicKey: PublicKey = keyFactory.generatePublic(publicKeySpec)

            val claims: Claims =
                Jwts.parserBuilder().setSigningKey(publicKey).build().parseClaimsJws(identityToken).body
            val subject = claims.subject
            val email = claims["email"] as String?
            return ProviderIdAndEmail(subject, email)
        } catch (e: Exception) {
            throw BookandException(ErrorCode.APPLE_LOGIN_ERROR)
        }
    }

    fun setHeaders(accessToken: String): HttpHeaders {
        val headers = HttpHeaders()
        headers["Authorization"] = "Bearer $accessToken"
        headers.contentType = MediaType.APPLICATION_FORM_URLENCODED
        headers.accept = listOf(MediaType.APPLICATION_JSON)
        return headers
    }

    fun getAccountByEmail(email: String): Account {
        return accountRepository.findByEmail(email)
            ?: throw BookandException(ErrorCode.NOT_FOUND_MEMBER)
    }

    fun getSuspendedAccount(account: Account): SuspendedAccount? {
        return suspendedAccountRepository.findByAccount(account)
    }

    private companion object {
        const val INIT_BOOKMARK_FOLDER_NAME = "모아보기"
    }
}